src/ICS20Transfer.sol

```diff
@@ -116,17 +117,14 @@ contract ICS20Transfer is IIBCApp, IICS20Transfer, IICS20Errors, Ownable, Reentr
             return ICS20Lib.errorAck(abi.encodePacked("invalid receiver: ", packetData.receiver));
         }
 
-        // TODO: Implement escrow balance tracking (#6)
-        if (originatorChainIsSource) {
+        if (receiverChainIsSource) {
+            address ics20EscrowAddress = _getEscrowAddress(msg_.packet.destChannel);
+            _transferFrom(ics20EscrowAddress, receiver, erc20Address, packetData.amount);
+        } else {
             // sender is source, so we mint vouchers
-            // NOTE: getReceiveTokenContractAndSource has already created a contract if it didn't exist
-            IBCERC20(erc20Address).mint(packetData.amount);
+            IBCERC20(erc20Address).mint(receiver, packetData.amount);
         }
```

Comment:
When receiving tokens: if the chain is the source of the tokens, then unscrew and transfer to the receiver; otherwise, mint tokens to the receiver on the ERC20 contract for the corresponding denomination.